internal/pkg/pkg.go

```diff
@@ -37,25 +52,25 @@ type Pkg struct {
 
 	// A package can contain zero or one Kptfile meta resource.
 	// A nil value represents an implicit package.
-	kptfile       *kptfilev1alpha2.KptFile
-	kptfileLoaded bool
-
-	// A package can contain zero or one Pipeline meta resource.
-	pipeline       *kptfilev1alpha2.Pipeline
-	pipelineLoaded bool
+	kptfile *kptfilev1alpha2.KptFile
 }
 
 // New returns a pkg given an absolute or relative OS-defined path.
 // Use ReadKptfile or ReadPipeline on the return value to read meta resources from filesystem.
 func New(path string) (*Pkg, error) {
 	p := filepath.Clean(path)
 
-	u, err := filepath.EvalSymlinks(p)
+	// Note(droot): Calling EvalSymlinks before Abs causes a strange issue on OSX
```

Comment:
So did some digging:
On my laptop (OSX), `/tmp` and `/var` are symlinks pointing to `/private/tmp` and `/private/var` respectively and luckily, I ran into this issue during a manual test. So if `EvalSymlinks` is called on a path that contains a symlinks component for ex. calling `EvalSymlinks(`/tmp/some-dir`)` will resolve to `/private/tmp/some-dir` and eventually `Abs(`/private/tmp/some-dir`)` resulting in `/private/tmp/some-dir`).  Versus, assuming you are under `/tmp` dir, calling `EvalSymlinks(`some-dir`)` ==> `some-dir` and Abs(`some-dir`) resulting in `/tmp/some-dir`. 

The above scenario is not a problem if you don't have nested package hierarchy, this becomes a problem in a recursive hydration because subpackages will have absolute paths `/tmp/some-dir/subpkg-a` and they will resolve to `/private/tmp/some-dir/subpkg-a` resulting in an inconsistent filesystem hierarchy for the pkg.

So changing the order and calling `Abs(..)` first forces the symlinks to be resolved always (or calling Abs on the rootpkg before calling pkg.New() can avoid the issue).

Now, I expect to see same behavior on Linux as well, but haven't verified it.

Reply:
Updated the comment with more details.